src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/FileBasedPrograms/VirtualProjectXmlProvider.cs

```diff
@@ -70,7 +86,15 @@ internal class VirtualProjectXmlProvider(DotnetCliHelper dotnetCliHelper)
 
             if (response is RunApiOutput.Project project)
             {
-
+                using (await _gate.DisposableWaitAsync(cancellationToken))
+                {
+                    _diagnosticsByFilePath.TryGetValue(documentFilePath, out var previousCachedDiagnostics);
+                    _diagnosticsByFilePath[documentFilePath] = project.Diagnostics;
+
+                    // check for difference, and signal to host to update if so.
+                    if (previousCachedDiagnostics.IsDefault || !project.Diagnostics.SequenceEqual(previousCachedDiagnostics))
```

Comment:
How do we remove diagnostics?  It looks really bad when we have stale diagnostics, so we definitely need to ensure we're clearing them when the file based project is deleted / renamed / otherwise changes contexts?

Reply:
We also probably need to remove them from the `_diagnosticsByFilePath` map if the file/project is deleted, otherwise we have a small leak

Reply:
I'm wrapping this method to try and ensure that the cache is always dealt with appropriately, whatever the response from run-api.